### ISSUE TITLE
DebugContext - error files gets overriden on outline scenarios

### ISSUE THREAD

**Issue by jorgetutor** [Metadrop/behat-contexts#130]
When a step fails a screenshot is automatically generated using the feature + step name. The file is overridden on the next outline variable so we only have the last executed scenario (attaching screenshot).

![image](https://user-images.githubusercontent.com/1230268/158592342-3ed61572-49a9-49c8-abc7-47f974b81cae.png)

Options:

- Improve naming to include outline variables.
- Add a random string to the file name. It will create several files between different executions, which I don't know if will be an issue or a feature.

**Comment by mistermoper** [comment]
I think using outline variables, which should be possible as we already have scenario information, is a good option. because it would help to classify the files. Maybe we should not let the outline variable values to exceed a specific length, in case there are variables with long values (currently it doesn't happen).

I think having different files for different locations may produce a lot of garbage, and I don't know a clear reason to need to preserve files from older executions.

**Comment by rsanzante** [comment]
I agree with @mistermoper on adding a outline variable to the files.

However, I think it would be interesting to keep different results. Sometimes you are in a debugging session an is interesting to check the evolution. For this I would create a new folder for each execution. But this is a complete new issue (having different folders or each test run).

**Comment by jorgetutor** [comment]
The file name is calculated in the nuvoleweb library. Should we patch/contribute it or override it as we extend it?
`
  /**
   * Take a screenshot after failed steps or save the HTML for non js drivers.
   *
   * @AfterStep
   */
  public function takeScreenshotAfterFailedStep(AfterStepScope $event) {
    if ($event->getTestResult()->getResultCode() !== TestResult::FAILED) {
      // Not a failed step.
      return;
    }
    try {
      $step = $event->getStep();
      $file_name = str_replace('/', '-', $event->getFeature()->getFile()) . '-' . $step->getKeyword() . '_' . $step->getText();
      $file_name = 'behat-failed__' . $file_name;
      $message = "Screenshot for failed step created in @file_name";
      $this->createScreenshotsForErrors($file_name, $message, $event->getTestResult());
    }
    catch (DriverException $e) {
    }
  }
`

**Comment by omarlopesino** [comment]
Probably we can propose a PR to nuvole/drupal-behat overriding the filename in the YML configuration.

Something like this:

```
filename: behat-failed__[feature]-[step-keyword]-[step-text]
```

That would be the default name.

Then, if we want to override it in other projects we can put

```
filename: behat-failed__[feature]-[outline-identifier]-[step-keyword]-[step-text]
```

Then code is not needed to override the filename, only configuration.

What do you think @rsanzante @jorgetutor ?

**Comment by rsanzante** [comment]
Are those tokens already available or we should add code to support them?

**Comment by omarlopesino** [comment]
Both the tokens and the filename YAML property do not exist, it must be added code in the ScreenshotContext to support them. I was thinking in doing a basic str_replace

**Comment by rsanzante** [comment]
For what I'm see in the [original code](https://github.com/nuvoleweb/drupal-behat/blob/72de2ee0f419d8882316fb39e1fcfc3795527f41/src/DrupalExtension/Context/ScreenShotContext.php#L129) only the outline identifier needs to be added. 

The problem is [AfterStepScope](https://github.com/Behat/Behat/blob/master/src/Behat/Behat/Hook/Scope/AfterStepScope.php) has no knowledge of the outline. This means we would have create our own outline identifier on each scenario outline  (I guess  
 using [BeforeScenarioScope](https://github.com/Behat/Behat/blob/master/src/Behat/Behat/Hook/Scope/BeforeScenarioScope.php) and [AfterScenarioScope](https://github.com/Behat/Behat/blob/master/src/Behat/Behat/Hook/Scope/AfterScenarioScope.php).

But this seems too complex.

What about checking the filename before writing and adding a number (_1, _2, _3) if it already exists?
